src/include/sof/audio/buffer.h

```diff
@@ -209,32 +209,27 @@ bool buffer_params_match(struct comp_buffer __sparse_cache *buffer,
 static inline void buffer_stream_invalidate(struct comp_buffer __sparse_cache *buffer,
 					    uint32_t bytes)
 {
-	if (!is_coherent_shared(buffer, c))
-		return;
-
-	audio_stream_invalidate(&buffer->stream, bytes);
+	if (buffer->is_shared)
+		audio_stream_invalidate(&buffer->stream, bytes);
```

Comment:
what about DMA buffers - on the host or DAI edge? Are they also handled by this class, if yes - their caches have to be managed, don't they?

Reply:
and they are in the very same way as before
This PR does not touch the audio data, just control structures
and in 100% of situations where it was "is_coherent_shared" now is "is_shared". +

For DAIs and DMA buffers the structures were **not** marked as shared:  coherent_shared_thread for buffers were called only once in the whole code - in comp_buffer_connect, called from ipc helpers (possible to mark buffer as shared)  and from IPC probes (not possible to mark as shared)

so cache operations were not performed at this point. And as CI is green - it is still working as DMA won't work at all using cached alias

It is, however, a very good question and I'll investigate it

Reply:
both ask and answer are clear, see below mails information in June:

On Tue, 6 Jun 2023, Tian, Baofeng wrote:

> Thanks, Guennadi, this is aligned.
>
> I am not sure whether close source have similar invalidate/writeback implementation or not, I don't remember old FW have this implementation.
> I don't know too much history about SOF, any particular reason to add invalidate and writeback for each buffer operation?
> Normally, it should be automatic(on ARM or old FW), only need add in some special case.

You have to synchronise cache for DMA.

Macin's implementation is quite clear and easy to understand, Guennadi's concern is dma buffer.
it need writeback or invalidate before sending/get DMA, question is: is it really required?

Reply:
> and they are in the very same way as before
> This PR does not touch the audio data, just control structures

Not sure this is correct. Look at `audio_stream_invalidate()` - it synchronises audio data, not control structures. And before it was done always, now you move it to only be done for shared buffers

Reply:
@lyakh 
no. it was not, look like removed code:

```
static inline void buffer_stream_invalidate(struct comp_buffer __sparse_cache *buffer,
					    uint32_t bytes)
{
	if (!is_coherent_shared(buffer, c))
		return;

	audio_stream_invalidate(&buffer->stream, bytes);
}
```

Reply:
indeed, you're right, sorry for the noise!